lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -34,421 +25,546 @@
+            continue;
+          // Check if sinkPort is a reaching def of sourcePort.
+          auto reachingDefsToSrcPort =
+              modulePaths->second.find(modSrcPortField);
+          if (reachingDefsToSrcPort != modulePaths->second.end())
+            if (llvm::find(reachingDefsToSrcPort->second, modSinkPortField) !=
+                reachingDefsToSrcPort->second.end()) {
+              // This case can occur when there are multiple RWProbes on the
+              // port, which refer to the same base value. So, each of such
+              // probes are reaching defs of each other. Hence the false
+              // loops. Instead of recording this in the reachingDefGraph,
+              // record it separately with the rwProbeClasses.
+              setOfEquivalentRWProbes.insert(srcNode);
+              if (minArgNum > srcArgNum) {
+                // Make one of the RWProbe port the base node. Use the first
+                // port for deterministic error messages.
+                minArgNum = srcArgNum;
+                basePortNode = srcNode;
+              }
+              continue;
+            }
         }
-        FieldRef instanceOutPort(inst.getResult(outPortNum),
-                                 modOutPort.getFieldID());
-        llvm::append_range(children, fieldToVals[instanceOutPort]);
+        addReachingDef(sinkPort, srcPort);
       }
-    } else if (auto mem = dyn_cast<MemOp>(ref.getDefiningOp())) {
-      if (mem.getReadLatency() > 0)
+      if (setOfEquivalentRWProbes.empty())
+        continue;
+
+      // Add all the rwprobes to the same class.
+      for (auto probe : setOfEquivalentRWProbes)
+        rwProbeClasses.unionSets(probe, sinkNode);
+
+      // Make the first port as the base value.
+      // Note: this is a port and the actual reference base exists in another
+      // module.
+      auto leader = rwProbeClasses.getLeaderValue(sinkNode);
+      rwProbeRefersTo[leader] = basePortNode;
+
+      setOfEquivalentRWProbes.insert(sinkNode);
+      // Add a reaching def  from base RWProbe port to all other RWProbe ports.
+      for (auto probe : setOfEquivalentRWProbes)
+        if (probe != basePortNode)
+          reachingDefGraph[probe].second.push_back(basePortNode);
+    }
+  }
+
+  // Record the FieldRef, corresponding to the result of the sub op
+  // `result = base[index]`
+  void recordValueRefersToFieldRef(Value base, unsigned fieldID, Value result) {
+
+    // Check if base is itself field of an aggregate.
+    auto it = valToFieldRefs.find(base);
+    if (it != valToFieldRefs.end()) {
+      // Rebase it to the original aggregate.
+      // Because of subaccess op, each value can refer to multiple FieldRefs.
+      SmallVector<FieldRef> entry;
+      for (auto &sub : it->second)
+        entry.emplace_back(sub.getValue(), sub.getFieldID() + fieldID);
+      // Update the map at the end, to avoid invaliding the iterator.
+      valToFieldRefs[result].append(entry.begin(), entry.end());
```

Comment:
As part of this analysis, this computes the set of all fieldref's a value may point to (may be more than one due to subaccess).  I'm not sure how likely this is to reach this pass in practice (or due to pipeline behavior), but consider:

```mlir
firrtl.circuit "Vec" {
  firrtl.module @Vec(in %x: !firrtl.vector<vector<uint<1>, 65536>, 65536>, in %sel: !firrtl.uint<16>, out %y: !firrtl.vector<vector<uint<1>, 65536>, 65536>) {
    %c_x, %c_sel, %c_y = firrtl.instance c interesting_name @Child(in x: !firrtl.vector<vector<uint<1>, 65536>, 65536>, in sel: !firrtl.uint<16>, out y: !firrtl.vector<vector<uint<1>, 65536>, 65536>)
    firrtl.strictconnect %c_x, %x : !firrtl.vector<vector<uint<1>, 65536>, 65536>
    firrtl.strictconnect %c_sel, %sel : !firrtl.uint<16>
    firrtl.strictconnect %y, %c_y : !firrtl.vector<vector<uint<1>, 65536>, 65536>
  }
  firrtl.module private @Child(in %x: !firrtl.vector<vector<uint<1>, 65536>, 65536>, in %sel: !firrtl.uint<16>, out %y: !firrtl.vector<vector<uint<1>, 65536>, 65536>) {
    %0 = firrtl.subaccess %y[%sel] : !firrtl.vector<vector<uint<1>, 65536>, 65536>, !firrtl.uint<16>
    %1 = firrtl.subaccess %0[%sel] : !firrtl.vector<uint<1>, 65536>, !firrtl.uint<16>
    %2 = firrtl.subaccess %x[%sel] : !firrtl.vector<vector<uint<1>, 65536>, 65536>, !firrtl.uint<16>
    %3 = firrtl.subaccess %2[%sel] : !firrtl.vector<uint<1>, 65536>, !firrtl.uint<16>
    firrtl.strictconnect %1, %3 : !firrtl.uint<1>
  }
}
```

Which if fed through `circt-opt --firrtl-check-comb-loops` with (or without) this PR really bogs down (warning: I killed it as it neared running out of memory locally).

FWIW `LowerTypes` also struggles with this input (recently learned it tries to convert subaccess operations even when preserving aggregates!) so I'll let you decide if this is relevant-- just wanted to demonstrate the issue.

If this needs addressing there's still good argument that this is much nicer + faster + does more than what we have now so don't mean to gate keep on doing ALL the things, just trying to avoid surprises as we start preserving aggregates (and subaccesses) that may reach this pass.